warehouse/query-core/src/main/java/datawave/query/jexl/visitors/PushdownMissingIndexRangeNodesVisitor.java

```diff
@@ -72,78 +74,43 @@ public PushdownMissingIndexRangeNodesVisitor(ShardQueryConfiguration config, Met
                 JexlNode newChild = (JexlNode) node.jjtGetChild(i).jjtAccept(this, data);
                 andNode.jjtAddChild(newChild, i);
                 newChild.jjtSetParent(andNode);
             }
+            return andNode;
         }
-        
-        return andNode;
     }
     
     /**
      * Delay the ranges that overlap holes. The range map is expected to only be indexed ranges.
      */
-    protected JexlNode delayIndexBoundedRange(Map<LiteralRange<?>,List<JexlNode>> ranges, List<JexlNode> leaves, ASTAndNode currentNode, JexlNode newNode,
-                    Object data) {
-        // Add all children in this AND/OR which are not a part of the range
-        JexlNodes.ensureCapacity(newNode, leaves.size() + ranges.size());
-        int index = 0;
-        for (; index < leaves.size(); index++) {
-            log.debug(leaves.get(index).image);
-            // Add each child which is not a part of the bounded range, visiting them first
-            JexlNode visitedChild = (JexlNode) leaves.get(index).jjtAccept(this, null);
-            newNode.jjtAddChild(visitedChild, index);
-            visitedChild.jjtSetParent(newNode);
-        }
-        
-        for (Map.Entry<LiteralRange<?>,List<JexlNode>> range : ranges.entrySet()) {
-            // If we have any terms that we expanded, wrap them in parens and add them to the parent
-            ASTAndNode rangeNodes = new ASTAndNode(ParserTreeConstants.JJTANDNODE);
-            
-            JexlNodes.ensureCapacity(rangeNodes, range.getValue().size());
-            for (int i = 0; i < range.getValue().size(); i++) {
-                rangeNodes.jjtAddChild(range.getValue().get(i), i);
-            }
-            
-            JexlNode child = rangeNodes;
-            if (missingIndexRange(range.getKey())) {
-                child = IndexHoleMarkerJexlNode.create(rangeNodes);
-            } else {
-                child = JexlNodes.wrap(rangeNodes);
-            }
-            
-            newNode.jjtAddChild(child, index++);
-            child.jjtSetParent(newNode);
-        }
+    protected JexlNode delayBoundedIndexHole(LiteralRange range, ASTAndNode currentNode, Object data) {
         
-        // If we had no other nodes than this bounded range, we can strip out the original parent
-        if (newNode.jjtGetNumChildren() == 1) {
-            newNode.jjtGetChild(0).jjtSetParent(newNode.jjtGetParent());
-            return newNode.jjtGetChild(0);
+        if (missingIndexRange(range)) {
+            return IndexHoleMarkerJexlNode.create(currentNode);
+        } else {
+            return currentNode;
         }
         
-        return newNode;
     }
     
     @Override
     public Object visit(ASTReferenceExpression node, Object data) {
         // if not already delayed somehow
-        if (!QueryPropertyMarker.instanceOf(node, null)) {
+        if (!QueryPropertyMarkerVisitor.instanceOfAnyExcept(node, Collections.singletonList(BoundedRange.class))) {
```

Comment:
Previously, QueryPropertyMarkerVisitor.instanceOf(node, null) implied that we were searching for any delayed nodes.  Now that we have BoundedRange as an additional marker, it implies something different as that isn't delayed.  Are we sure that there aren't any other places in the code where we don't need to do a similar update to what you've done here to ensure that BoundedRange isn't treated as a delayed node?  Should we create some kind of isInstanceOfDelayed method to make this clearer?

Reply:
Agreed.  I did validate all calls as such.